### PR TITLE
[WIP][SPARK-47321][INFRA] Upgrade `dawidd6/action-download-artifact` to v4 and `scacap/action-surefire-report` to latest

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to upgrade
- `dawidd6/action-download-artifact` from `09385b76de790122f4da9c82b17bccf858b9557c` to `v3.1.2`
- `scacap/action-surefire-report` from `482f012643ed0560e23ef605a79e8e87ca081648` to `v1.8.0`.

### Why are the changes needed?
- Eliminate warning messages about the workflow `test_report` running, especially the deprecated issue regarding `node16`.
  https://github.com/apache/spark/actions/runs/8184857052
  image

- dawidd6/action-download-artifact
  https://github.com/dawidd6/action-download-artifact/releases/tag/v3.0.0
  image

- scacap/action-surefire-report
  https://github.com/ScaCap/action-surefire-report/releases/tag/v1.8.0
  image

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Observe the running status of the workflow `test_report` in github.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: .github/workflows/test_report.yml [hunk 1]
```diff
@@ -30,14 +30,14 @@ jobs:
     runs-on: ubuntu-latest
     steps:
     - name: Download test results to report
-      uses: dawidd6/action-download-artifact@09385b76de790122f4da9c82b17bccf858b9557c # pin@v2
```
Comment (HyukjinKwon) [inline .github/workflows/test_report.yml:33]: The specific commit was set here because it might contain insecure codes (it's a personal repository).

Comment (panbingkun) [reply .github/workflows/test_report.yml:33]: So we should `pin` to a `specific commit` instead of using something like `v1`, right?

Comment (HyukjinKwon) [reply .github/workflows/test_report.yml:33]: Yeah, let's pin the commit version after reviewing if their codes have anything harmful.

Comment (panbingkun) [reply .github/workflows/test_report.yml:33]: Okay, let me review its source code very carefully.

**Pull request closed without merging**